File: backend/common/view_visibility.py
"""Canonical view-visibility tiers — the backend's single source of truth.

Mirrors ``frontend/src/types/viewVisibility.ts``. The two files must stay
in lock-step: the DB CHECK constraints, the API validators and the access
evaluator all reference these values.

Prior to this module the three-value enum was duplicated as bare string
literals in five places (both CHECK constraints in ``db/models.py``, the
validator in ``endpoints/views.py``, and the branch chain in
``services/view_access.py``), with a further ~15 copies on the frontend
that had already drifted apart in wording and spelling. Adding a tier
meant a wide, silent-bug-prone sweep; now every caller imports from here.

The tiers form a **strictly widening ladder** — each reaches everyone the
one above it reaches, plus more:

===========  ==========================================================
``private``  creator + explicit ``resource_grants`` + workspace admins
             of the view's workspace
``workspace``+ anyone holding ``workspace:view:read`` in that workspace
``enterprise``+ anyone holding a binding in *any* workspace
``public``   + *any* authenticated account, including one with no
             workspace bindings at all
===========  ==========================================================

Why ``enterprise`` means "holds a binding somewhere": this codebase has
no ``organizations`` table — workspaces are the only tenancy unit, and
"org" exists solely as a role tier (``system:org-admin``). "Has at least
one workspace binding" is the honest proxy for organisation membership,
and it is computable from ``PermissionClaims.ws_perms`` with no query.
Without that distinction ``enterprise`` and ``public`` would be the same
tier — which is what ``enterprise`` degenerated to before this change,
where it was implemented as a bare "is authenticated" check.

``public`` is deliberately **authenticated-only**. There is no anonymous
or logged-out access to any view, and no share-link tokens.
"""
from __future__ import annotations

from enum import Enum


class ViewVisibility(str, Enum):
    """Visibility tier for a saved view.

    ``str`` mixin so existing string comparisons keep working —
    ``ViewVisibility.PRIVATE == "private"`` is True. Use
    ``ViewVisibility.<X>.value`` only when you need the bare string.
    """

    #: Creator, explicit grants, and workspace admins of its workspace.
    PRIVATE = "private"
    #: Anyone with read access in the view's own workspace.
    WORKSPACE = "workspace"
    #: Anyone with a binding in any workspace ("in the organisation").
    ENTERPRISE = "enterprise"
    #: Any authenticated account, workspace binding or not.
    PUBLIC = "public"


#: Tuple (not frozenset) because ordering matters for the DB CHECK
#: constraint text and for rendering the tier ladder in API errors.
VISIBILITY_VALUES: tuple[str, ...] = tuple(v.value for v in ViewVisibility)

#: The default for a newly created view, and the fallback for any legacy
#: row with a NULL ``visibility``. Fails safe — narrowest tier.
DEFAULT_VISIBILITY: str = ViewVisibility.PRIVATE.value


def visibility_check_constraint(column: str = "visibility") -> str:
    """SQL fragment for the ``CHECK`` constraint on a visibility column.

    Used by ``db/models.py`` for both ``views`` and ``context_models`` and
    by the alembic revision that widens the constraint, so the three can
    never disagree about the permitted set.
    """
    values = ", ".join(f"'{v}'" for v in VISIBILITY_VALUES)
    return f"{column} IN ({values})"


def is_valid_visibility(value: object) -> bool:
    """True when ``value`` is one of the permitted tiers."""
    return isinstance(value, str) and value in VISIBILITY_VALUES


__all__ = [
    "ViewVisibility",
    "VISIBILITY_VALUES",
    "DEFAULT_VISIBILITY",
    "visibility_check_constraint",
    "is_valid_visibility",
]

File: frontend/src/types/viewVisibility.ts
/**
 * Canonical view-visibility tiers — the single source of truth for the FE.
 *
 * Prior to this module the union `'private' | 'workspace' | 'enterprise'`
 * was redeclared in ~15 files and the option lists (label, description,
 * icon) were duplicated across five selectors and four read-only badge
 * maps. They had already drifted: the same tier was described as
 * "Only you can see this view" in the create wizard, "Only you (and
 * workspace admins) can see this view" in the share dialog, and "Only you
 * can see it" in the edit panel — with `organisation`/`organization`
 * spelled both ways. Every consumer now imports from here.
 *
 * The backend's equivalent lives in `backend/common/view_visibility.py` —
 * the two files must stay in lock-step. That module owns the DB CHECK
 * constraint and the access evaluator; this one owns how the tiers are
 * named and described to users.
 */
import { Building2, Globe, Lock, Users } from 'lucide-react'
import type { LucideIcon } from 'lucide-react'

export const VIEW_VISIBILITY = {
    PRIVATE: 'private',
    WORKSPACE: 'workspace',
    ENTERPRISE: 'enterprise',
    PUBLIC: 'public',
} as const

export type ViewVisibility = (typeof VIEW_VISIBILITY)[keyof typeof VIEW_VISIBILITY]

/**
 * Ordered narrowest → widest. The tiers form a strictly widening ladder:
 * each reaches everyone the previous one reaches, plus more. Render them
 * in this order everywhere so the escalation is visually obvious.
 */
export const VISIBILITY_ORDER: readonly ViewVisibility[] = [
    VIEW_VISIBILITY.PRIVATE,
    VIEW_VISIBILITY.WORKSPACE,
    VIEW_VISIBILITY.ENTERPRISE,
    VIEW_VISIBILITY.PUBLIC,
] as const

export interface VisibilityMeta {
    /** Short label for badges, chips and menu items. */
    label: string
    /** One-line explanation of exactly who this reaches. */
    description: string
    /** Sentence used when confirming a change to this tier. */
    reach: string
    icon: LucideIcon
    /** Tailwind colour family for accent styling. */
    accent: 'slate' | 'sky' | 'indigo' | 'amber'
}

export const VISIBILITY_META: Record<ViewVisibility, VisibilityMeta> = {
    private: {
        label: 'Private',
        description: 'Only you, people you share it with, and workspace admins',
        reach: 'only you, anyone you share it with explicitly, and admins of its workspace',
        icon: Lock,
        accent: 'slate',
    },
    workspace: {
        label: 'Workspace',
        description: 'Everyone in this view’s workspace',
        reach: 'everyone with access to this view’s workspace',
        icon: Users,
        accent: 'sky',
    },
    enterprise: {
        label: 'Enterprise',
        description: 'Everyone in the organisation, in any workspace',
        reach: 'everyone in the organisation — including people who are not in this workspace',
        icon: Building2,
        accent: 'indigo',
    },
    public: {
        label: 'Public',
        description: 'Everyone signed in to the platform, workspace or not',
        reach: 'every signed-in account on the platform, including people with no workspace access at all',
        icon: Globe,
        accent: 'amber',
    },
}

/** Fallback for a legacy or unrecognised value read off the wire. */
export const DEFAULT_VISIBILITY: ViewVisibility = VIEW_VISIBILITY.PRIVATE

export function isViewVisibility(value: unknown): value is ViewVisibility {
    return typeof value === 'string'
        && (VISIBILITY_ORDER as readonly string[]).includes(value)
}

/**
 * Meta for a value that may be legacy, absent, or a tier this build
 * doesn't know about. Never throws — the badge must always render.
 */
export function visibilityMeta(value: unknown): VisibilityMeta {
    return VISIBILITY_META[isViewVisibility(value) ? value : DEFAULT_VISIBILITY]
}

/**
 * True when moving `from` → `to` exposes the view to more people. Drives
 * the confirmation step on bulk changes and the warning copy in the share
 * dialog: narrowing is safe, widening deserves a beat of friction.
 */
export function isWidening(from: unknown, to: unknown): boolean {
    const a = VISIBILITY_ORDER.indexOf(isViewVisibility(from) ? from : DEFAULT_VISIBILITY)
    const b = VISIBILITY_ORDER.indexOf(isViewVisibility(to) ? to : DEFAULT_VISIBILITY)
    return b > a
}

/**
 * Why a given caller can see a given view. Mirrors
 * `backend/app/services/view_access.py: ReadReason`, stamped per-caller on
 * `View.grantedBy`.
 *
 * This exists because nothing in the product could answer "why can I see
 * this view?" — which is why a visibility leak took a code audit to
 * diagnose rather than a glance at a tooltip.
 */
export const GRANTED_BY_COPY: Record<string, string> = {
    'creator': 'You created this view',
    'grant': 'Shared with you directly',
    'workspace-admin': 'You administer this view’s workspace',
    'system-admin': 'You’re a platform administrator',
    'tier:workspace': 'Shared with this view’s workspace',
    'tier:enterprise': 'Shared with the whole organisation',
    'tier:public': 'Shared with everyone on the platform',
}

/** Human sentence for a `grantedBy` value, or null when it's absent. */
export function grantedByCopy(reason: unknown): string | null {
    if (typeof reason !== 'string') return null
    return GRANTED_BY_COPY[reason] ?? null
}
